Feature#481 resources needed tab

marvin-bitterlich opened the pull request:
Closes #481

# Changes proposed in this pull request:
Open the resources needed tab in the builder's hut gui:

Have list of the needed resources with added informations: each item have it name display, the number of items missing for the builder (i.e how many are needed minus the one in his inventory - the chest's inventory), the number of items available and the number needed. Example:
- Cobblestone -2  8/10 (needed 10, have 8, need 2 more)
- Oak wood  -10 0/10 (needed 10 , have none, need 10 more)
- Oak wood plank +90  100/10 (need 10, have 100, have 90 to many )

Click on a resource needed item to add it to the chest from the player inventory
Additionally the resources are colour coded:
- Black: the builder have enough
- Red: the builder will need more of them
- Green: the builder will need more of them but the player have the correct amount in his inventory.

The list is ordered as follow:
- Green first;
- Red with button enabled (i.e. the player can give some of those items but not all of them)
- Red with button disabled (i.e. the player does not have any in his inventory)
- Black with button disabled (i.e. the builder have enough)

The list is refreshed when:
- The player is opening the gui.
- The player give some resources from the gui
- The chest or NPC inventory change
- The list is *not* refreshed when one of the additional chest have changed
-  Refresh the list every time we click on an item to reflect the changes

Review please

xavierh committed:
Add the amount of available resources from the chest and the builder inventory next to the
amount needed

Resources are color in red if the builder will need more of them

xavierh committed:
new method InventoryUtils.setOverSizedStack which act as setStack but allow to pass a "too big" stack

xavierh committed:
Merge branch 'version/1.11' of https://github.com/Minecolonies/minecolonies into feature#481-resources-needed-tab

xavierh committed:
Allow to transfert items from the player inventory to the builder chest within the resources needed tab

xavierh committed:
Fix the update for the Builder gui
remove logging

xavierh committed:
4 colors in the resource needed tab:
* black: no need
* green: player have enough in his inventory
* orange: player have some but not enough
* red: the player does not have this resource

xavierh committed:
set the color to black if no other color

xavierh committed:
Fix color green (when in player inventory)
Better ordering

xavierh committed:
* Fix setOwerSizeStack
* Use damage
* Replace orange color by red + enable button
* Disable button for previous red state (No item available)

xavierh committed:
Merge branch 'version/1.11' of https://github.com/Minecolonies/minecolonies into feature#481-resources-needed-tab

Conflicts:
	src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java

xavierh committed:
add support for multiple chest when calculating available resources

I still need to trigger the update whenever an additional chest's content is modified.
Add add the ability to drop in additionnal chest

xavierh committed:
Add resources to the additional chest whne using the gui

xavierh committed:
Mark the builderbuilding dirty when opening its gui to make sure we get up to date view

xavierh committed:
Mark the BuilderBuilding dirty when the delivery man transfert items to one of its chest

xavierh committed:
Fix doc

xavierh committed:
fix more javadoc

xavierh committed:
Mark the building as dirty whenever the delivery man drop some item into its chests

xavierh committed:
Serialize the ItemStack myself as sometime it does give bad result
Left the log in for the time being

xavierh committed:
Logging

xavierh committed:
Remove logging

xavierh committed:
cleanup

xavierh committed:
Merge branch 'version/1.11' of https://github.com/Minecolonies/minecolonies into feature#481-resources-needed-tab

Conflicts:
	src/main/java/com/minecolonies/coremod/MineColonies.java

xavierh committed:
Display how many item are needed instead of the number of item already there

i.e. instead of:
Oak Wood 10/40
it will show -30/40

xavierh committed:
remove blank lines

xavierh committed:
Removed unused import

xavierh committed:
remove unused import

xavierh committed:
fix javadoc comment

xavierh committed:
Fix spelling

xavierh committed:
Remove orange color as it is not used

xavierh committed:
Merge branch 'version/1.11' of https://github.com/Minecolonies/minecolonies into feature#481-resources-needed-tab

Conflicts:
	src/main/java/com/minecolonies/coremod/client/gui/WindowHutBuilder.java
	src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java
	src/main/java/com/minecolonies/coremod/inventory/InventoryCitizen.java

xavierh committed:
Tidy up imports

xavierh committed:
Display the number of missing items AND the number available

marvin-bitterlich committed:
Merge pull request #528 from xavierh/feature#481-resources-needed-tab

Feature#481 resources needed tab

marvin-bitterlich commented:
@xavierh please fill out http://minecolonies.com/recruitment and we can add you to the team :) then you can push to in house branches

xavierh commented:
Cheers. Will do later on today

On 30 Jan 2017 16:54, "Marvin Hofmann" <notifications@github.com> wrote:

> @xavierh <https://github.com/xavierh> please fill out
> http://minecolonies.com/recruitment and we can add you to the team :)
> then you can push to in house branches
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Minecolonies/minecolonies/pull/529#issuecomment-276118327>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACplvNvm2FI4i89m1pzCSl3-anfcO-Msks5rXhXJgaJpZM4LxpCm>
> .
>

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java:
```diff
@@ -339,6 +339,15 @@ public int getLevel()
         return level;
     }
 
+    public void onInventoryChanged()
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Document this public method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUndocumentedApi)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/network/messages/TransferItemsRequestMessage.java:
```diff
@@ -0,0 +1,154 @@
+        }
+
+        final Item item = message.itemStack.getItem();
+        final int amountInPlayer = InventoryUtils.getItemCountInInventory(player.inventory, item, message.itemStack.getMetadata());
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Remove this unused "amountInPlayer" local variable. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1481)
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Remove this useless assignment to local variable "amountInPlayer". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1854)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/network/messages/TransferItemsRequestMessage.java:
```diff
@@ -0,0 +1,154 @@
+import org.jetbrains.annotations.NotNull;
+
+
+/**
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') First sentence should end with a period. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=checkstyle%3AJavadoc_Style)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/network/messages/TransferItemsRequestMessage.java:
```diff
@@ -0,0 +1,154 @@
+     * The id of the colony.
+     */
+    private int      colonyId;
+    /**
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') First sentence should end with a period. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=checkstyle%3AJavadoc_Style)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/network/messages/TransferItemsRequestMessage.java:
```diff
@@ -0,0 +1,154 @@
+        final Item item = message.itemStack.getItem();
+        final int amountInPlayer = InventoryUtils.getItemCountInInventory(player.inventory, item, message.itemStack.getMetadata());
+        final int amountToTake = Math.min(message.quantity, InventoryUtils.getItemCountInInventory(player.inventory, item, message.itemStack.getItemDamage()));
+        ItemStack itemStackToTake = new ItemStack(item, amountToTake, message.itemStack.getItemDamage());
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Local variable 'itemStackToTake' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/network/messages/TransferItemsRequestMessage.java:
```diff
@@ -0,0 +1,154 @@
+     * How many item need to be transfer from the player inventory to the building chest
+     */
+    private ItemStack      itemStack;
+    /**
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') First sentence should end with a period. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=checkstyle%3AJavadoc_Style)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/network/messages/TransferItemsRequestMessage.java:
```diff
@@ -0,0 +1,154 @@
+
+        ItemStack remainingItemStack = InventoryUtils.setOverSizedStack(building.getTileEntity(), itemStackToTake);
+
+        if (remainingItemStack.getCount()>0) //we still have some to drop, let's try the additional chests now
```
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png 'Severity: INFO') Move this trailing comment on the previous empty line. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3ATrailingCommentCheck)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/client/gui/WindowHutBuilder.java:
```diff
@@ -64,9 +124,8 @@ public String getBuildingName()
+        pullResourcesFromHut();
 
         final ScrollingList resourceList = findPaneOfTypeByID(LIST_RESOURCES, ScrollingList.class);
         resourceList.setDataProvider(new ScrollingList.DataProvider()
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Reduce this anonymous class number of lines from 48 to at most 30, or make it a named class. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1188)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/client/gui/WindowHutBuilder.java:
```diff
@@ -35,16 +65,46 @@ public WindowHutBuilder(final BuildingBuilder.View building)
+                            return resource1.getName().compareTo(resource2.getName());
+                        }
+    
+                        return (resource2.getAvailabilityStatus().compareTo(resource1.getAvailabilityStatus()));
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Remove those useless parentheses. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUselessParenthesesCheck)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/client/gui/WindowHutBuilder.java:
```diff
@@ -18,12 +29,31 @@
+    private static final String RESOURCE_QUANTITY_MISSING          = "resourceQuantity";
+    private static final int RESOURCE_QUANTITY_MISSING_POSITION    = 4;
+
+    private static final int red       = Color.getByName("red",0);
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Rename this constant name to match the regular expression '^[A-Z][A-Z0-9]*(_[A-Z0-9]+)*$'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS00115)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/client/gui/WindowHutBuilder.java:
```diff
@@ -35,16 +65,46 @@ public WindowHutBuilder(final BuildingBuilder.View building)
     {
-        if (builder.getColony().getBuilding(builder.getID()) != null)
+        final AbstractBuilding.View newView = builder.getColony().getBuilding(builder.getID());
+        if (newView != null && newView instanceof BuildingBuilder.View)
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') No need to check for null before an instanceof [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ASimplifyConditional)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/client/gui/WindowHutBuilder.java:
```diff
@@ -18,12 +29,31 @@
+    private static final int RESOURCE_QUANTITY_MISSING_POSITION    = 4;
+
+    private static final int red       = Color.getByName("red",0);
+    private static final int darkgreen = Color.getByName("darkgreen",0);
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Rename this constant name to match the regular expression '^[A-Z][A-Z0-9]*(_[A-Z0-9]+)*$'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS00115)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/client/gui/WindowHutBuilder.java:
```diff
@@ -18,12 +29,31 @@
+
+    private static final int red       = Color.getByName("red",0);
+    private static final int darkgreen = Color.getByName("darkgreen",0);
+    private static final int black     = Color.getByName("black",0);
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Rename this constant name to match the regular expression '^[A-Z][A-Z0-9]*(_[A-Z0-9]+)*$'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS00115)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/client/gui/WindowHutBuilder.java:
```diff
@@ -76,21 +135,57 @@ public int getElementCount()
             {
-                final Object obj = entries[index];
+                    final BuildingBuilder.BuildingBuilderResource resource = resources.get(index);
+                    Label resourceLabel = rowPane.findPaneOfTypeByID(RESOURCE_NAME, Label.class);
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Local variable 'resourceLabel' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/client/gui/WindowHutBuilder.java:
```diff
@@ -76,21 +135,57 @@ public int getElementCount()
-                final Object obj = entries[index];
+                    final BuildingBuilder.BuildingBuilderResource resource = resources.get(index);
+                    Label resourceLabel = rowPane.findPaneOfTypeByID(RESOURCE_NAME, Label.class);
+                    Label neededLabel = rowPane.findPaneOfTypeByID(RESOURCE_AVAILABLE_NEEDED, Label.class);
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Local variable 'neededLabel' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/client/gui/WindowHutBuilder.java:
```diff
@@ -2,13 +2,24 @@
+import com.minecolonies.coremod.network.messages.MarkBuildingDirtyMessage;
+import com.minecolonies.coremod.network.messages.TransferItemsRequestMessage;
+import com.minecolonies.coremod.util.InventoryUtils;
+import com.minecolonies.coremod.util.Log;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Remove this unused import 'com.minecolonies.coremod.util.Log'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/client/gui/WindowHutBuilder.java:
```diff
@@ -76,21 +135,57 @@ public int getElementCount()
+                    final BuildingBuilder.BuildingBuilderResource resource = resources.get(index);
+                    Label resourceLabel = rowPane.findPaneOfTypeByID(RESOURCE_NAME, Label.class);
+                    Label neededLabel = rowPane.findPaneOfTypeByID(RESOURCE_AVAILABLE_NEEDED, Label.class);
+                    Button addButton = rowPane.findPaneOfTypeByID(RESOURCE_ADD, Button.class);
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Local variable 'addButton' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/client/gui/WindowHutBuilder.java:
```diff
@@ -76,21 +135,57 @@ public int getElementCount()
+                    addButton.setPosition(buttonX,buttonY);
+
+                    resourceLabel.setLabelText(resource.getName());
+                    final int missing = (resource.getAvailable()-resource.getNeeded());
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Remove those useless parentheses. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUselessParenthesesCheck)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/client/gui/WindowHutBuilder.java:
```diff
@@ -76,21 +135,57 @@ public int getElementCount()
+
+                    resourceLabel.setLabelText(resource.getName());
+                    final int missing = (resource.getAvailable()-resource.getNeeded());
+                    final String missingItems = (missing>0)?"+"+Integer.toString(missing):Integer.toString(missing);
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Add parentheses to make the operator precedence explicit. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS864)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/client/gui/WindowHutBuilder.java:
```diff
@@ -103,4 +198,24 @@ public void onUpdate()
+        @NotNull final ItemStack itemStack = resources.get(index).getItemStack();
+        @NotNull final Label quantityLabel = (Label) button.getParent().getChildren().get(RESOURCE_QUANTITY_MISSING_POSITION);
+        final int quantity = Integer.parseInt(quantityLabel.getLabelText());
+        final String buttonLabel = button.getID();
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Remove this unused "buttonLabel" local variable. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1481)
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Remove this useless assignment to local variable "buttonLabel". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1854)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/client/gui/WindowHutBuilder.java:
```diff
@@ -35,16 +65,46 @@ public WindowHutBuilder(final BuildingBuilder.View building)
+                resource.setPlayerAmount(InventoryUtils.getItemCountInInventory(inventory, item, resource.getItemStack().getItemDamage()));
+            }
+
+            Collections.sort(resources, new Comparator<BuildingBuilder.BuildingBuilderResource>() 
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Make this anonymous inner class a lambda [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1604)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:
```diff
@@ -165,6 +176,95 @@ public void writeToNBT(@NotNull final NBTTagCompound compound)
+        }
+
+        private String name;
+        private ItemStack itemStack;
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Private field 'itemStack' could be made final; it is only initialized in the declaration or constructor. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3AImmutableField)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:
```diff
@@ -165,6 +176,95 @@ public void writeToNBT(@NotNull final NBTTagCompound compound)
+        private String name;
+        private ItemStack itemStack;
+        private int amountAvailable;
+        private int amountNeeded;
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Private field 'amountNeeded' could be made final; it is only initialized in the declaration or constructor. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3AImmutableField)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:
```diff
@@ -246,6 +352,60 @@ public void resetNeededResources()
     }
 
     /**
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') First sentence should end with a period. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=checkstyle%3AJavadoc_Style)

MinecoloniesSonar replied:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') First sentence should end with a period. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=checkstyle%3AJavadoc_Style)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:
```diff
@@ -165,6 +176,95 @@ public void writeToNBT(@NotNull final NBTTagCompound compound)
+            return RessourceAvailability.NOT_NEEDED;
+        }
+
+        public String toString()
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Add the "@Override" annotation above this method signature [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1161)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:
```diff
@@ -246,6 +352,60 @@ public void resetNeededResources()
+            final IInventory chestInventory = this.getTileEntity();
+            if (chestInventory!=null)
+            {
+                resource.setAvailable(resource.getAvailable() + InventoryUtils.getItemCountInInventory(chestInventory, entry.getValue().getItem(), entry.getValue().getItemDamage()));
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Split this 182 characters long line (which is greater than 180 authorized). [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS00103)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:
```diff
@@ -246,6 +352,60 @@ public void resetNeededResources()
+                    final TileEntity entity = builder.world.getTileEntity(pos);
+                    if(entity instanceof TileEntityChest)
+                    {
+                        resource.setAvailable(resource.getAvailable() + InventoryUtils.getItemCountInInventory((TileEntityChest)entity, entry.getValue().getItem(), entry.getValue().getItemDamage()));
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Split this 199 characters long line (which is greater than 180 authorized). [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS00103)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:
```diff
@@ -165,6 +176,95 @@ public void writeToNBT(@NotNull final NBTTagCompound compound)
     }
 
     /**
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') First sentence should end with a period. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=checkstyle%3AJavadoc_Style)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:
```diff
@@ -165,6 +176,95 @@ public void writeToNBT(@NotNull final NBTTagCompound compound)
     /**
+     * Hold the pair of number of avaliable and needed StackItem
+     */
+    public static class BuildingBuilderResource
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Reduce this class from 73 to the maximum allowed 50 or externalize it in a public class. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS2972)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:
```diff
@@ -246,6 +352,60 @@ public void resetNeededResources()
+
+        for (@NotNull final Map.Entry<String, ItemStack> entry : neededResources.entrySet())
+        {
+            ItemStack itemStack = entry.getValue();
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Local variable 'itemStack' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:
```diff
@@ -165,6 +176,95 @@ public void writeToNBT(@NotNull final NBTTagCompound compound)
+     */
+    public static class BuildingBuilderResource
+    {
+        public enum RessourceAvailability
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Document this public enum. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUndocumentedApi)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:
```diff
@@ -45,6 +52,10 @@
+    /**
+     * Contains all resources available in the chest or builder inventory needed for a certain build.
+     */
+    private HashMap<String, BuildingBuilder.BuildingBuilderResource> resourcesAvailable = new HashMap<>();
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Private field 'resourcesAvailable' could be made final; it is only initialized in the declaration or constructor. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3AImmutableField)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:
```diff
@@ -246,6 +352,60 @@ public void resetNeededResources()
+
+            if (builderInventory!=null)
+            {
+                resource.setAvailable(resource.getAvailable() + InventoryUtils.getItemCountInInventory(builderInventory, entry.getValue().getItem(), entry.getValue().getItemDamage()));
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Split this 184 characters long line (which is greater than 180 authorized). [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS00103)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:
```diff
@@ -165,6 +176,95 @@ public void writeToNBT(@NotNull final NBTTagCompound compound)
+            HAVE_ENOUGHT
+        }
+
+        private String name;
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Private field 'name' could be made final; it is only initialized in the declaration or constructor. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3AImmutableField)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/network/messages/MarkBuildingDirtyMessage.java:
```diff
@@ -0,0 +1,86 @@
+import org.jetbrains.annotations.NotNull;
+import com.minecolonies.coremod.util.Log;
+
+/**
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') First sentence should end with a period. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=checkstyle%3AJavadoc_Style)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/util/InventoryUtils.java:
```diff
@@ -507,6 +512,32 @@ public static ItemStack setStack(@NotNull final IInventory inventory, @Nullable
+            stackSize-=itemCount;
+            final ItemStack remainingItems = setStack(inventory,items);
+            stackSize+=remainingItems.getCount();
+            if (items.getCount()==remainingItems.getCount())
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Missing curly brace. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS00121)

xavierh commented:
Does it mean I have commit right on the new branch ?
Will sonarqube still provide its help ?

marvin-bitterlich commented:
@xavierh on your local repo just change the `origin` upstream to the minecolonies repo and use the same branch. You can now commit here

xavierh committed:
Fix sonar

xavierh committed:
Merge branch 'feature#481-resources-needed-tab' of https://github.com/minecolonies/minecolonies into feature#481-resources-needed-tab

MinecoloniesSonar commented:
SonarQube analysis reported 22 issues
* ![MAJOR][MAJOR] 11 major
* ![MINOR][MINOR] 11 minor

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [BuildingBuilder.java#L414](https://github.com/Minecolonies/minecolonies/blob/a3a0f7923cfcd53d877e2eced4a240b897d47a04/src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java#L414): Document this public method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUndocumentedApi)
1. ![MINOR][MINOR] [TransferItemsRequestMessage.java](https://github.com/Minecolonies/minecolonies/blob/a3a0f7923cfcd53d877e2eced4a240b897d47a04/src/main/java/com/minecolonies/coremod/network/messages/TransferItemsRequestMessage.java): Replace all tab characters in this file by sequences of white-spaces. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS00105)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

xavierh committed:
Move BuidlingBuilder.View to BuildingBuilderView
more sonar fixes

xavierh committed:
More sonar fixes

xavierh committed:
Merge branch 'version/1.11' of https://github.com/Minecolonies/minecolonies into feature#481-resources-needed-tab

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/client/gui/WindowHutBuilder.java:
```diff
@@ -76,21 +137,61 @@ public int getElementCount()
+        addButton.setPosition(buttonX,buttonY);
+
+        resourceLabel.setLabelText(resource.getName());
+        final int missing = (resource.getAvailable()-resource.getNeeded());
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Remove those useless parentheses. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUselessParenthesesCheck)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/client/gui/WindowHutBuilder.java:
```diff
@@ -76,21 +137,61 @@ public int getElementCount()
+
+        resourceLabel.setLabelText(resource.getName());
+        final int missing = (resource.getAvailable()-resource.getNeeded());
+        final String missingItems = (missing>0)?"+"+Integer.toString(missing):Integer.toString(missing);
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Add parentheses to make the operator precedence explicit. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS864)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:
```diff
@@ -165,6 +173,106 @@ public void writeToNBT(@NotNull final NBTTagCompound compound)
+            HAVE_ENOUGHT
+        }
+
+        final private String name;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Reorder the modifiers to comply with the Java Language Specification. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AModifiersOrderCheck)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:
```diff
@@ -1,26 +1,30 @@
-import com.minecolonies.blockout.views.Window;
 import com.minecolonies.coremod.achievements.ModAchievements;
-import com.minecolonies.coremod.client.gui.WindowHutBuilder;
+import com.minecolonies.coremod.colony.buildings.BuildingBuilderView;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Remove this unused import 'com.minecolonies.coremod.colony.buildings.BuildingBuilderView'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:
```diff
@@ -165,6 +173,106 @@ public void writeToNBT(@NotNull final NBTTagCompound compound)
+        }
+
+        final private String name;
+        final private ItemStack itemStack;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Reorder the modifiers to comply with the Java Language Specification. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AModifiersOrderCheck)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:
```diff
@@ -165,6 +173,106 @@ public void writeToNBT(@NotNull final NBTTagCompound compound)
+     * - How many are available to the builder
+     * - How many are in the player's inventory (client side only)
+     */
+    public static class BuildingBuilderResource
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Reduce this class from 74 to the maximum allowed 50 or externalize it in a public class. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS2972)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:
```diff
@@ -165,6 +173,106 @@ public void writeToNBT(@NotNull final NBTTagCompound compound)
+        final private String name;
+        final private ItemStack itemStack;
+        private int amountAvailable;
+        private int amountNeeded;
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Private field 'amountNeeded' could be made final; it is only initialized in the declaration or constructor. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3AImmutableField)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:
```diff
@@ -246,84 +360,91 @@ public void resetNeededResources()
-            neededResources = new HashMap<>();
+            if (builderInventory!=null)
+            {
+                resource.setAvailable(resource.getAvailable() + InventoryUtils.getItemCountInInventory(builderInventory, entry.getValue().getItem(), entry.getValue().getItemDamage()));
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Split this 184 characters long line (which is greater than 180 authorized). [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS00103)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:
```diff
@@ -246,84 +360,91 @@ public void resetNeededResources()
+     * @param stack the stack to test.
+     * @return true if so.
+     */
+    public boolean requiresResourceForBuilding(ItemStack stack)
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Document this public method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUndocumentedApi)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:
```diff
@@ -246,84 +360,91 @@ public void resetNeededResources()
     }
 
     /**
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') First sentence should end with a period. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=checkstyle%3AJavadoc_Style)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:
```diff
@@ -246,84 +360,91 @@ public void resetNeededResources()
-                final String block = ByteBufUtils.readUTF8String(buf);
-                final int amount = buf.readInt();
-                neededResources.put(block, amount);
+                resource.setAvailable(resource.getAvailable() + InventoryUtils.getItemCountInInventory(chestInventory, entry.getValue().getItem(), entry.getValue().getItemDamage()));
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Split this 182 characters long line (which is greater than 180 authorized). [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS00103)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilderView.java:
```diff
@@ -0,0 +1,95 @@
+            final ItemStack itemStack = new ItemStack(Item.getByNameOrId(Integer.toString(itemId)),1,damage);
+            final int amountAvailable = buf.readInt();
+            final int amountNeeded = buf.readInt();
+            final BuildingBuilder.BuildingBuilderResource resource = new BuildingBuilder.BuildingBuilderResource(itemStack.getDisplayName(), itemStack, amountAvailable,amountNeeded);
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Split this 182 characters long line (which is greater than 180 authorized). [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS00103)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilderView.java:
```diff
@@ -0,0 +1,95 @@
+import com.minecolonies.coremod.colony.ColonyView;
+import com.minecolonies.coremod.util.Utils;
+import io.netty.buffer.ByteBuf;
+import net.minecraft.nbt.NBTTagList;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Remove this unused import 'net.minecraft.nbt.NBTTagList'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

MinecoloniesSonar commented:
SonarQube analysis reported 16 issues
* ![MAJOR][MAJOR] 6 major
* ![MINOR][MINOR] 10 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

xavierh committed:
More sonar fixes

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:
```diff
@@ -165,6 +172,106 @@ public void writeToNBT(@NotNull final NBTTagCompound compound)
+     */
+    public static class BuildingBuilderResource
+    {
+        final private String name;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Reorder the modifiers to comply with the Java Language Specification. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AModifiersOrderCheck)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:
```diff
@@ -165,6 +172,106 @@ public void writeToNBT(@NotNull final NBTTagCompound compound)
+    public static class BuildingBuilderResource
+    {
+        final private String name;
+        final private ItemStack itemStack;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Reorder the modifiers to comply with the Java Language Specification. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AModifiersOrderCheck)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:
```diff
@@ -165,6 +172,106 @@ public void writeToNBT(@NotNull final NBTTagCompound compound)
+    {
+        final private String name;
+        final private ItemStack itemStack;
+        final private int amountNeeded;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Reorder the modifiers to comply with the Java Language Specification. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AModifiersOrderCheck)

MinecoloniesSonar commented:
SonarQube analysis reported 8 issues
* ![MAJOR][MAJOR] 3 major
* ![MINOR][MINOR] 5 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

xavierh committed:
Fix sonar issues

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:
```diff
@@ -165,6 +172,106 @@ public void writeToNBT(@NotNull final NBTTagCompound compound)
+
+        private int amountAvailable;
+        private int amountPlayer;
+        final private String name;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Reorder the modifiers to comply with the Java Language Specification. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AModifiersOrderCheck)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:
```diff
@@ -165,6 +172,106 @@ public void writeToNBT(@NotNull final NBTTagCompound compound)
+        private int amountAvailable;
+        private int amountPlayer;
+        final private String name;
+        final private ItemStack itemStack;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Reorder the modifiers to comply with the Java Language Specification. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AModifiersOrderCheck)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:
```diff
@@ -165,6 +172,106 @@ public void writeToNBT(@NotNull final NBTTagCompound compound)
+        private int amountPlayer;
+        final private String name;
+        final private ItemStack itemStack;
+        final private int amountNeeded;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Reorder the modifiers to comply with the Java Language Specification. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AModifiersOrderCheck)

xavierh commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:
```diff
@@ -165,6 +172,106 @@ public void writeToNBT(@NotNull final NBTTagCompound compound)
+        private int amountAvailable;
+        private int amountPlayer;
+        final private String name;
+        final private ItemStack itemStack;
```
Can somebody tell me what I am doing wrong here ?

cltnschlosser replied:
private final

cltnschlosser replied:
In general you have Ex: `private static final String`, or `Visibility [static] [final] Type`

xavierh replied:
Cheers, I thought I would go crazy

xavierh committed:
final-y :)

MinecoloniesSonar commented:
SonarQube analysis reported 5 issues
* ![MAJOR][MAJOR] 3 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

xavierh committed:
more sonar fixes

MinecoloniesSonar commented:
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

Raycoms commented on src/main/java/com/minecolonies/coremod/client/gui/WindowHutBuilder.java:
```diff
@@ -76,21 +137,67 @@ public int getElementCount()
+                resourceLabel.setColor(RED, RED);
+                neededLabel.setColor(RED, RED);
+                break;
+            case HAVE_ENOUGHT:
```
I think it should be HAVE_ENOUGH

Raycoms commented on src/main/java/com/minecolonies/coremod/client/gui/WindowHutBuilder.java:
```diff
@@ -76,21 +137,67 @@ public int getElementCount()
+
+        resourceLabel.setLabelText(resource.getName());
+        final int missing = resource.getAvailable()-resource.getNeeded();
+        final String missingItems = (missing>0)?("+"+Integer.toString(missing)):Integer.toString(missing);
```
can you leave some space to increase readability before and after "+" "-" "?" ">" ":"

Raycoms commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:
```diff
@@ -246,84 +360,91 @@ public void resetNeededResources()
     }
 
     /**
```
Information about a resource. (He wants a period after the first sentence in each javadoc.)

xavierh replied:
Unless I am looking at the wrong line, the dot is here with no space behind

Raycoms commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:
```diff
@@ -165,6 +172,106 @@ public void writeToNBT(@NotNull final NBTTagCompound compound)
+        private final ItemStack itemStack;
+        private final int amountNeeded;
+
+        BuildingBuilderResource(final String name, final ItemStack itemStack, final int amountAvailable, final int amountNeeded)
```
Please declare visibility.

Raycoms commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:
```diff
@@ -165,6 +172,106 @@ public void writeToNBT(@NotNull final NBTTagCompound compound)
+
+        public RessourceAvailability getAvailabilityStatus()
+        {
+            if (amountNeeded >amountAvailable)
```
also space after ">" and before and after "==" "<" "-"

Raycoms commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilderView.java:
```diff
@@ -0,0 +1,95 @@
+package com.minecolonies.coremod.colony.buildings;
```
can you create a new package in the buildings package "buildingViews"?

Raycoms reviewed (commented):
Seems nice overally.
Only some things about readability.

xavierh committed:
Fix sonar with anonymous class

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:
```diff
@@ -165,6 +172,131 @@ public void writeToNBT(@NotNull final NBTTagCompound compound)
+         *
+         * This is use in the gui to order the list of resources needed.
+         */
+        public static class ResourceComparator implements Comparator<BuildingBuilderResource>
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Make this class "Serializable". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS2063)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:
```diff
@@ -165,6 +172,131 @@ public void writeToNBT(@NotNull final NBTTagCompound compound)
+         */
+        public static class ResourceComparator implements Comparator<BuildingBuilderResource>
+        {
+            /**
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') First sentence should end with a period. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=checkstyle%3AJavadoc_Style)

MinecoloniesSonar commented:
SonarQube analysis reported 4 issues
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

xavierh committed:
move BuildingBuilderView to a new package buildingviews
more sonar fixes

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:
```diff
@@ -165,6 +172,131 @@ public void writeToNBT(@NotNull final NBTTagCompound compound)
+        private final ItemStack itemStack;
+        private final int amountNeeded;
+
+        public BuildingBuilderResource(final String name, final ItemStack itemStack, final int amountAvailable, final int amountNeeded)
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Document this public constructor. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUndocumentedApi)

MinecoloniesSonar commented:
SonarQube analysis reported 5 issues
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [buildingviews](https://github.com/Minecolonies/minecolonies/blob/3e10269026b8a93eb3b399f1a85072493681d714/src/main/java/com/minecolonies/coremod/colony/buildings/buildingviews): Add a 'package-info.java' file to document the 'buildingviews' package [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1228)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

xavierh committed:
more sonar fixes
+ address some recommendation from Raycoms

xavierh committed:
missing import

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/buildingviews/package-info.java:
```diff
@@ -0,0 +1,4 @@
+/**
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') First sentence should end with a period. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=checkstyle%3AJavadoc_Style)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/buildingviews/package-info.java:
```diff
@@ -0,0 +1,4 @@
+/**
+ * Contains all the view for the buildings
+ */
+package com.minecolonies.coremod.colony.building.buildingviews;
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') This file "package-info.java" should be located in "com/minecolonies/coremod/colony/building/buildingviews" directory, not in "/home/travis/build/Minecolonies/minecolonies/src/main/java/com/minecolonies/coremod/colony/buildings/buildingviews". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1598)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:
```diff
@@ -165,6 +173,139 @@ public void writeToNBT(@NotNull final NBTTagCompound compound)
+         *
+         * This is use in the gui to order the list of resources needed.
+         */
+        public static class ResourceComparator implements Comparator<BuildingBuilderResource>, Serializable
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Add a "static final long serialVersionUID" field to this class. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS2057)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:
```diff
@@ -165,6 +173,139 @@ public void writeToNBT(@NotNull final NBTTagCompound compound)
+         */
+        public static class ResourceComparator implements Comparator<BuildingBuilderResource>, Serializable
+        {
+            /**
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') First sentence should end with a period. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=checkstyle%3AJavadoc_Style)

MinecoloniesSonar commented:
SonarQube analysis reported 5 issues
* ![MAJOR][MAJOR] 3 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

xavierh committed:
Fix package name

xavierh committed:
sonar

MinecoloniesSonar commented:
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

cltnschlosser committed:
Here goes nothing, attempt to externalize class, without rebuilding my environment for this minecraft version

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/buildingviews/BuildingBuilderView.java:
```diff
@@ -0,0 +1,96 @@
+import com.minecolonies.blockout.views.Window;
+import com.minecolonies.coremod.client.gui.WindowHutBuilder;
+import com.minecolonies.coremod.colony.buildings.AbstractBuildingWorker;
+import com.minecolonies.coremod.colony.buildings.BuildingBuilder;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Remove this unused import 'com.minecolonies.coremod.colony.buildings.BuildingBuilder'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:
```diff
@@ -1,26 +1,30 @@
-import java.util.HashMap;
-import java.util.Map;
+import java.util.*;
+import java.io.Serializable;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Remove this unused import 'java.io.Serializable'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

MinecoloniesSonar commented:
SonarQube analysis reported 2 issues
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

cltnschlosser committed:
Make externalized views possible (#537)

* Make externalized views possible

* Add final keyword

xavierh committed:
Use only one list for the resources needed and available
Use StorageItem as a base class for BuildingBuilderResource

xavierh committed:
fix setAmount javadoc

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilderResource.java:
```diff
@@ -0,0 +1,162 @@
+ * - How many are available to the builder
+ * - How many are in the player's inventory (client side only)
+ */
+public class BuildingBuilderResource extends ItemStorage
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Override this superclass' "equals" method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS2160)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:
```diff
@@ -246,84 +261,84 @@ public void resetNeededResources()
+        for (@NotNull final Map.Entry<String, BuildingBuilderResource> entry : neededResources.entrySet())
         {
-            super.deserialize(buf);
+            BuildingBuilderResource resource = entry.getValue();
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR') Local variable 'resource' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

MinecoloniesSonar commented:
SonarQube analysis reported 4 issues
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [BuildingBuilder.java](https://github.com/Minecolonies/minecolonies/blob/04d287d395aed55b2e613bc2ead428598780c552/src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java): Replace all tab characters in this file by sequences of white-spaces. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS00105)

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:
```diff
@@ -201,13 +213,16 @@ public void serializeToView(@NotNull ByteBuf buf)
-        int preAmount = 0;
-        if (this.neededResources.containsKey(res.getUnlocalizedName()))
+	BuildingBuilderResource resource = this.neededResources.get(res.getUnlocalizedName());
+        if (resource != null)
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Avoid if (x != y) ..; else ..; [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3AConfusingTernary)

cltnschlosser commented:
```
FAILURE: Build failed with an exception.
* What went wrong:
Could not resolve all dependencies for configuration ':forgeGradleMcDeps'.
> Could not download lwjgl.jar (org.lwjgl.lwjgl:lwjgl:2.9.4-nightly-20150209)
   > Could not get resource 'https://libraries.minecraft.net/org/lwjgl/lwjgl/lwjgl/2.9.4-nightly-20150209/lwjgl-2.9.4-nightly-20150209.jar'.
      > Could not GET 'https://libraries.minecraft.net/org/lwjgl/lwjgl/lwjgl/2.9.4-nightly-20150209/lwjgl-2.9.4-nightly-20150209.jar'. Received status code 500 from server: Internal Server Error
```

Doesn't look like build failure is your fault

xavierh committed:
Follow sonar recomendation

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilderResource.java:
```diff
@@ -0,0 +1,175 @@
+    }
+
+    @Override
+    public boolean equals(final Object o)
```
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png 'Severity: BLOCKER') This class overrides "equals()" and should therefore also override "hashCode()". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1206)

MinecoloniesSonar commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilderResource.java:
```diff
@@ -0,0 +1,175 @@
+        {
+            final BuildingBuilderResource that = (BuildingBuilderResource) o;
+
+            return (this.getAvailable() == that.getAvailable() && this.getPlayerAmount() == that.getPlayerAmount());
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Remove those useless parentheses. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUselessParenthesesCheck)

MinecoloniesSonar commented:
SonarQube analysis reported 3 issues
* ![BLOCKER][BLOCKER] 1 blocker
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [BuildingBuilder.java](https://github.com/Minecolonies/minecolonies/blob/36d2ef0865ccc53b029663ce81c5f64e1b9cbb2d/src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java): Replace all tab characters in this file by sequences of white-spaces. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS00105)

[BLOCKER]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png 'Severity: BLOCKER'
[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

xavierh committed:
hashCode for BuildingBuilderResource
+ other sonar

cltnschlosser commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilderResource.java:
```diff
@@ -138,12 +138,18 @@ public boolean equals(final Object o)
+    @Override
+    public int hashCode()
+    {
+        return super.hashCode() + amountAvailable + amountPlayer;
```
Not great, but someone else can generate one later.

xavierh replied:
However I am curious to know why?

cltnschlosser replied:
Because hashcodes should be unique-ish and shouldn't collide very often. Anytime amountAvailable + amountPlayer equal the same thing you will have a hash collision. There have been plenty of studies on optimial hash functions, and one of the simplest ones is just multiplying everything by 31 at each step, so in this case you would do:

31*(31*super.hashCode() + amountAvailable) + amountPlayer


Sorry for any formatting issues, on mobile.

xavierh commented:
However I am curious to know why?

xavierh committed:
move BuildingBuilderRessource to a new package
Gui:
- Change possition of the button to not cover the text
- Show how much more of one resource the player need (only when he does not have enough)

xavierh commented:
I do that then. Cheers

On 1 Feb 2017 17:59, "Colton Schlosser" <notifications@github.com> wrote:

> *@cltnschlosser* commented on this pull request.
> ------------------------------
>
> In src/main/java/com/minecolonies/coremod/colony/buildings/
> BuildingBuilderResource.java
> <https://github.com/Minecolonies/minecolonies/pull/529>:
>
> >          }
>
>          return false;
>      }
>
> +    @Override
> +    public int hashCode()
> +    {
> +        return super.hashCode() + amountAvailable + amountPlayer;
>
> Because hashcodes should be unique-ish and shouldn't collide very often.
> Anytime amountAvailable + amountPlayer equal the same thing you will have a
> hash collision. There have been plenty of studies on optimial hash
> functions, and one of the simplest ones is just multiplying everything by
> 31 at each step, so in this case you would do:
>
> 31*(31*super.hashCode() + amountAvailable) + amountPlayer
>
> Sorry for any formatting issues, on mobile.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Minecolonies/minecolonies/pull/529>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/ACplvLEYxYNA0zAyFQ6kq_JshIsXpUD7ks5rYMgMgaJpZM4LxpCm>
> .
>

xavierh committed:
better hashCode

xavierh committed:
Fix label position

xavierh committed:
Fix issue with empty item stack or too big stack

xavierh committed:
remove logging

xavierh committed:
Merge branch 'version/1.11' of https://github.com/minecolonies/minecolonies into feature#481-resources-needed-tab

Conflicts:
	src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java

xavierh committed:
Missing commit

cltnschlosser commented on src/main/java/com/minecolonies/coremod/client/gui/WindowHutBuilder.java:
```diff
@@ -212,15 +214,22 @@ public void onUpdate()
-        MineColonies.getNetwork().sendToServer(new TransferItemsRequestMessage(this.building, itemStack, quantity));
+        if (res == null)
+        {
+            Log.getLogger().warn("WindowHutBuilder.transferItems: Error - Could not found the resource.");
```
could not find the resource

cltnschlosser commented on src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:
```diff
@@ -214,6 +215,11 @@ public void serializeToView(@NotNull ByteBuf buf)
      */
     public void addNeededResource(@Nullable final ItemStack res, final int amount)
     {
+        if (res.isEmpty() || (amount == 0))
```
those parenthesis are unneeded. The ones around the `amount == 0`

xavierh committed:
Remove unneeded parenthesis
Fix spelling

cltnschlosser commented:
LGTM, but has anyone tested it on a server?

[![Approved with PullApprove](https://img.shields.io/badge/maintainers-approved-brightgreen.svg)](https://pullapprove.com/Minecolonies/minecolonies/pull-request/529/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=Minecolonies/minecolonies)

xavierh commented:
I tested it on a small server i.e. max 2 concurrent players

xavierh committed:
Show correct resource name by taking into account the damage value

xavierh commented:
Not sure why it failed to build.

I will fix the javadoc warning getAvailable @return missing

xavierh committed:
Fix indentation
add @return for getAvailable

Raycoms commented:
:shipit: I'm okay with it.

[![Approved with PullApprove](https://img.shields.io/badge/maintainers-approved-brightgreen.svg)](https://pullapprove.com/Minecolonies/minecolonies/pull-request/529/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=Minecolonies/minecolonies)
